feat: add `poo/objets_calistheniques` (#14)
readthedocs-fr/notions: poo/objets_calistheniques/fr/OBJETS_CALISTHENIQUES.MD

> # Objets Calisthéniques

Reviewer: Préciser que cet article est une traduction d'un autre

        return redirect("homepage");
    }

    addFlash("error", "Bad credentials");

    return redirect("login");
}
```

Votre condition peut être **optimiste**, ce qui signifie que vous avez des conditions pour vos cas d'erreur et que le reste de votre méthode suit le scénario par défaut,
ou vous pouvez adopter une approche **défensive**, ce qui signifie que vous mettez le scénario par défaut dans une condition, et s'il n'est pas satisfait,
vous renvoyez un état d'erreur.
C'est mieux car cela évite les problèmes potentiels auxquels vous n'avez pas pensé.

Comme autre alternative, vous pouvez également introduire une variable afin de rendre votre instruction return paramétrable. Bien que pratique, ce n'est cependant pas toujours possible.

```Java
public void login(String username, String password) {
    String redirectRoute = "homepage";

    if (!userRepository.isValid(username, password)) {
        addFlash("error", "Bad credentials");
        redirectRoute = "login";
    }

    redirect(redirectRoute);
}
```

En outre, il convient de mentionner que la programmation orientée objet nous offre des fonctionnalités **puissantes**, telles que le **polymorphisme**. 

> Enfin, les patterns **Null Object**, **State** et **Strategy** peuvent également vous aider !

Reviewer: suggestion : ajouter des liens vers des articles présentant ces patterns ?

## Règle 3 : Encapsulez tous les primitifs et chaîne de caractères

Suivre cette règle est assez simple, il vous suffit d'encapsuler tous les primitifs dans les objets, afin d'éviter l'anti-pattern Primitive Obsession.

Si la variable primitive a un comportement, vous DEVEZ l'encapsuler.
Pour cela, vous pouvez utiliser des **Value Objects** comme `Money` ou `Hour` par exemple. (Une fiche sur les value objets arrive bientôt)

## Règle 4 : Une collection par classe

Toute classe qui contient une collection ne doit contenir **aucun autre field**.
Si vous avez un ensemble d'éléments (une collection) et que vous souhaitez les manipuler, __créez une classe dédiée à cet ensemble__.

Chaque collection est **enveloppée dans sa propre classe**, donc maintenant tous les comportements liés à cette collection doivent être implémentés dans la classe contenant la collection (par exemple les méthodes de filtrage).

## Règle 5 : Un point par ligne (One dot per line)

Ce *point* est celui que vous utilisez pour **appeler des méthodes** en Java. Ce serait une flèche en PHP, mais qui utilise PHP de toute façon ? :D

> Fondamentalement, la règle dit que vous ne devez pas enchaîner les appels de méthode. 

Reviewer: ```suggestion
Fondamentalement, la règle dit que vous ne devez pas enchaîner les appels de méthodes. 
```

}
```
```Java
class Board {
    public String boardRepresentation() {
        StringBuilder builder = new StringBuilder();

        for (Location loc : squares()) {
            builder.append(loc.current.representation.substring(0, 1));
        }

        return builder.toString();
    }
}
```
Il est normal d'avoir des attributs publics dans `Piece` et `Location`.
En fait, avoir une propriété `publique` ou une propriété `private` avec getter/setter est la même chose (voir règle 9).

Cependant, la méthode `boardRepresentation()` est horrible, jetez un œil à cette ligne :

`builder.append(loc.current.representation.substring(0, 1));`

Il accède à une `Location`, puis à son field `piece`, puis au field `representation` de la référence de `Piece` sur laquelle il effectue une action.
C'est loin d'être un point par ligne.

Mais, [la loi de Déméter](http://wiki.c2.com/?LawOfDemeter) vous dit de parler uniquement à vos amis, alors faisons-le :
```Java
class Location {
    private Piece current;

    public void addTo(StringBuilder builder) {
        current.addTo(builder);
    }
}
```
Rendre privée le field `piece` garantit que vous n'essaierez pas de faire quelque chose de mal avec.

> Cependant, comme vous devez effectuer une action sur cee field, vous avez besoin d'une nouvelle méthode `addTo()`. 

Reviewer: ```suggestion
Cependant, comme vous devez effectuer une action sur ce field, vous avez besoin d'une nouvelle méthode `addTo()`. 
```

    }

    public void addTo(StringBuilder builder) {
        builder.append(character());
    }
}
```

Là encore, vous devez modifier la visibilité de votre attribut. 
Pour rappel, [le principe *ouvert/fermé*](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) dit que les entités logicielles (classes, modules, fonctions, etc.) doivent être ouvertes pour l'extension, mais fermées pour la modification.

De plus, extraire le code pour obtenir le premier caractère du representation dans une nouvelle méthode semble être une bonne idée car il peut être réutilisé à un moment donné. 
Enfin, voici la classe `Board` refactorisée :

```Java
class Board {
    public String boardRepresentation() {
        StringBuilder builder = new StringBuilder();

        for (Location location : squares()) {
            location.addTo(builder);
        }

        return builder.toString();
    }
}
```

Et voilà qui respecte cette règle et qui est en plus, plus lisible !

## Règle 6 : N'abrégez pas !

La bonne question étant : **pourquoi diable voulez-vous abréger ?**

Vous pouvez répondre que c'est parce que vous écrivez le même nom encore et encore ?

> Et je répondrais que cette méthode est réutilisée plusieurs fois, et qu'elle ressemble à une **duplication de code**.

Reviewer: ```suggestion
Et je répondrai que cette méthode est réutilisée plusieurs fois, et qu'elle ressemble à une **duplication de code**.
```
(je ne pas certain mais il me semble)

Vous direz donc ensuite que le nom de la méthode est de toute façon trop long. 

> Et je vous dirais que votre classe a peut-être de **multiples responsabilités**, ce qui est mauvais car cela viole le principe de [responsabilité unique](https://en.wikipedia.org/wiki/Single-responsibility_principle.

Reviewer: ```suggestion
Et je vous dirai que votre classe a peut-être de **multiples responsabilités**, ce qui est mauvais car cela viole le principe de [responsabilité unique](https://en.wikipedia.org/wiki/Single-responsibility_principle.
```
(je ne pas certain mais il me semble)

Je dis souvent que si vous ne trouvez pas un nom décent pour une classe ou une méthode, il y a probablement un problème. 
C'est une règle que j'utilise pour concevoir un logiciel en nommant des méthodes ou des attributs.

De plus si vous abrégez, la compréhension de votre code peut devenir ambiguë, ce qui n'est pas une bonne chose, même pour vous quand vous allez vous relire plusieurs semaines après avoir écrit ce code. 

> N'abbrégez, point final. 

Reviewer: ```suggestion
N'abbrégez pas, point final. 
```

## Règle 7 : Ne pas avoir d'entités trop grosses

Tout d'abord, par entité j'entends les packages, les classes, les méthodes, bref, tout ce qui peut être extensible.

Pas de classe sur 50 lignes et pas de package sur 10 fichiers.
Cela dépend de vous, mais je pense que vous pouvez changer le nombre de lignes de 50 à 150.

L'idée derrière cette règle est que les fichiers longs sont plus difficiles à lire, plus difficiles à comprendre et plus difficiles à gérer.
Donc plus vous avez des entités petites, plus elles sont faciles à comprendre et donc à maintenir, et donc plus vous gagnez de temps !

> ## Règle 8 : Pas de classes avec plus de deux instances de classe

Reviewer: ```suggestion
## Règle 8 : Pas de classe avec plus de deux attributs d'instance
```

Cette règle est probablement la plus difficile, mais elle favorise une cohésion élevée et une meilleure encapsulation dans votre classe.

> Il faut noter que cette règle est lié à la règle 3.

Reviewer: ```suggestion
Il faut noter que cette règle est liée à la règle 3.
```

## Règle 2 : N'utilisez pas le mot clé else

Le mot clé *else* est bien connu car il est intégré dans la construction *if/else* qui est présente dans presque tous les langages de programmation.
Nous sommes d'accord que lire des instructions conditionnelles imbriquées n'est pas agréable.
C'est exactement pourquoi il faut l'éviter. 
Il est beaucoup plus facile d'ajouter une nouvelle branche au code existant que de le réfactoriser vers une meilleure solution.
Et à force de faire cela, vous vous retrouvez souvent avec un très mauvais code qui est __illisible et incompréhensible__.

```Java
public void login(String username, String password) {
    if (userRepository.isValid(username, password)) {
        redirect("homepage");
    } else {
        addFlash("error", "Bad credentials");

        redirect("login");
    }
}
```

Cependant, il existe heureusement des solutions pour éviter ça, comme par exemple utliser le **retour anticipé** de la méthode.

```Java

> public void login(String username, String password) {

Reviewer 2: Hmm... Le type de retour de cette méthode est `void`, tu ne peux pas renvoyer de valeur

vous renvoyez un état d'erreur.
C'est mieux car cela évite les problèmes potentiels auxquels vous n'avez pas pensé.

Comme autre alternative, vous pouvez également introduire une variable afin de rendre votre instruction return paramétrable. Bien que pratique, ce n'est cependant pas toujours possible.

```Java
public void login(String username, String password) {
    String redirectRoute = "homepage";

    if (!userRepository.isValid(username, password)) {
        addFlash("error", "Bad credentials");
        redirectRoute = "login";
    }

    redirect(redirectRoute);
}
```

En outre, il convient de mentionner que la programmation orientée objet nous offre des fonctionnalités **puissantes**, telles que le **polymorphisme**. 
Enfin, les patterns [**Null Object**](https://en.wikipedia.org/wiki/Null_object_pattern), [**State**](https://en.wikipedia.org/wiki/State_pattern#:~:text=The%20state%20pattern%20is%20a,concept%20of%20finite%2Dstate%20machines.) et [**Strategy**](https://en.wikipedia.org/wiki/Strategy_pattern#:~:text=In%20computer%20programming%2C%20the%20strategy,family%20of%20algorithms%20to%20use.) peuvent également vous aider !

> ## Règle 3 : Encapsulez tous les primitifs et chaîne de caractères

Reviewer 2: ```suggestion
## Règle 3 : Encapsulez tous les primitifs et chaînes de caractères
```

Suivre cette règle est assez simple, il vous suffit d'encapsuler tous les primitifs dans les objets, afin d'éviter l'anti-pattern Primitive Obsession.

Si la variable primitive a un comportement, vous DEVEZ l'encapsuler.
Pour cela, vous pouvez utiliser des **Value Objects** comme `Money` ou `Hour` par exemple. (Une fiche sur les value objets arrive bientôt)

## Règle 4 : Une collection par classe

Toute classe qui contient une collection ne doit contenir **aucun autre field**.
Si vous avez un ensemble d'éléments (une collection) et que vous souhaitez les manipuler, __créez une classe dédiée à cet ensemble__.

Chaque collection est **enveloppée dans sa propre classe**, donc maintenant tous les comportements liés à cette collection doivent être implémentés dans la classe contenant la collection (par exemple les méthodes de filtrage).

Pour illustrer cette règle, voici un exemple provenant du code de la notion de [value objects](), nous avons une classe qui représente un ordre de commande :
```java

public class Order {

>     //Not a value objet because have an id

Reviewer 2: ```suggestion
    //Not a value object because it has an id
```

        this.products = products;
        this.addressToDeliver = addressToDeliver;
    }
}

```

Cette classe est composée d'un `UUID`, d'une `Address` et ce qui va nous intéresser ici d'un objet `Products` qui représente une liste de `Product`.

Jetons un oeil à cette classe :

```java
public class Products {

    private final List<Product> products;

    public Products() {
        this.products = new ArrayList<>();
    }

    public void addProduct(Product product) {
        this.products.add(product);
    }

    public Money getProductsCost() {
        Money money = new Money(0);
        for (Product product : products) {
            Money productCost = product.getCost();
            money = money.addMoney(productCost);
        }
        return money;
    }
}
```

Nous voyons bien que la classe ne comporte qu'un field qui est une collection de `Product`, la règle est donc respectée.
Nous pouvons donc observer que la logique pour gérer plusieurs `Product` est bien présente dans cette classe. 
On pourrait aussi facilement implémenter des méthodes de filtrage qui retourneraient tout les `Product` avec un prix minimum de 10€ par exemple.

> Maintenant, regardons comment se serait organisé le code s'il n'y avait pas de classe pour gérer cette collection, on aurait surement eu quelque chose de ce style : 

Reviewer 2: ```suggestion
Maintenant, regardons comment se serait organisé le code s'il n'y avait pas de classe pour gérer cette collection, on aurait sûrement eu quelque chose de ce style : 
```

```java
public class Order {

>     //Not a value objet because have an id

Reviewer 2: ```suggestion
    //Not a value object because it has an id
```

Author: Il faudrait aussi changer ça sur la pr des value objects, le code provenant de là-bas

    private final UUID orderId;
    private List<Product> products;
    private Address addressToDeliver;

    public Order(List<Product> products, Address addressToDeliver) {
        this.orderId = UUID.randomUUID();
        this.products = products;
        this.addressToDeliver = addressToDeliver;
    }

    public Money getProductsCost() {
        Money money = new Money(0);
        for (Product product : products) {
            Money productCost = product.getCost();
            money = money.addMoney(productCost);
        }
        return money;
    }
}
```

Nous voyons bien que ça aurait été à l'objet `Order` de gérer cette logique, alors qu'il aurait pus déléguer cela à une classe qui gère une collection de `Product`. 

> Pour l'instant ce n'est pas choquant, mais si on rajoute différentes méthodes de filtrages, la classe va vite commencer à s'allonger, ce qui ne respecterait pas la [règle 7](https://github.com/Emalios/notions/blob/calisthenics-objects-fr/poo/objets_calistheniques/fr/OBJETS_CALISTHENIQUES.MD#r%C3%A8gle-7--ne-pas-avoir-dentit%C3%A9s-trop-grosses) et serait très difficile à lire et maintenir.

Reviewer 2: ```suggestion
Pour l'instant ce n'est pas choquant, mais si on rajoute différentes méthodes de filtrage, la classe va vite commencer à s'allonger, ce qui ne respecterait pas la [règle 7](https://github.com/Emalios/notions/blob/calisthenics-objects-fr/poo/objets_calistheniques/fr/OBJETS_CALISTHENIQUES.MD#r%C3%A8gle-7--ne-pas-avoir-dentit%C3%A9s-trop-grosses) et serait très difficile à lire et maintenir.
```

## Règle 5 : Un point par ligne (One dot per line)

> Ce *point* est celui que vous utilisez pour **appeler des méthodes** en Java. Ce serait une flèche en PHP, mais qui utilise PHP de toute façon ? :D

Reviewer 2: Ce principe ne s'applique pas aux variables d'instances ?
this.myObj.someField.anotherOne respecte la règle ?

Author: Je pense oui, après de mon expérience c'est rare de voir ça

Fondamentalement, la règle dit que vous ne devez pas enchaîner les appels de méthodes. 
Cependant, cela ne s'applique pas aux [interfaces Fluent](http://flippinawesome.org/2013/05/20/fluent-apis-and-method-chaining/) et plus généralement à tout ce qui implémente le [modèle de chaînage de méthodes](https://martinfowler.com/dslCatalog/methodChaining.html) (par exemple, un générateur de requêtes ou un builder).

Pour les autres classes, vous devez respecter cette règle. C'est l'utilisation directe de [la loi de Déméter](http://wiki.c2.com/?LawOfDemeter), qui dit de ne parler **qu'à vos amis** et de ne pas parler à des étrangers.

Regardez ces classes :
```Java
class Location {
    public Piece current;
}
```
```Java
class Piece {
    public String representation;
}
```
```Java
class Board {
    public String boardRepresentation() {
        StringBuilder builder = new StringBuilder();

        for (Location loc : squares()) {
            builder.append(loc.current.representation.substring(0, 1));
        }

        return builder.toString();
    }
}
```
Il est normal d'avoir des attributs publics dans `Piece` et `Location`.

> En fait, avoir une propriété `publique` ou une propriété `private` avec getter/setter est la même chose (voir règle 9).

Reviewer 2: Cette affirmation est très controversée, et beaucoup de développeurs la considèrent comme fausse (moi inclus). Je suggérerais d'au moins préciser que c'est **selon les principes d'objets calisthéniques** que "attribut public == getter + setter", pour nuancer

Cependant, la méthode `boardRepresentation()` est horrible, jetez un œil à cette ligne :

`builder.append(loc.current.representation.substring(0, 1));`

Il accède à une `Location`, puis à son field `piece`, puis au field `representation` de la référence de `Piece` sur laquelle il effectue une action.
C'est loin d'être un point par ligne.

Mais, [la loi de Déméter](http://wiki.c2.com/?LawOfDemeter) vous dit de parler uniquement à vos amis, alors faisons-le :
```Java
class Location {
    private Piece current;

    public void addTo(StringBuilder builder) {
        current.addTo(builder);
    }
}
```

> Rendre privée le field `piece` garantit que vous n'essaierez pas de faire quelque chose de mal avec.

Reviewer 2: ```suggestion
Rendre privé le field `piece` garantit que vous n'essaierez pas de faire quelque chose de mal avec.
```

> Rendre privée le field `piece` garantit que vous n'essaierez pas de faire quelque chose de mal avec.

Reviewer 2: Je pense qu'il faudrait préciser ce que tu entends par "faire quelque chose de mal avec"

Reviewer 3: ```suggestion
Rendre privé le field `piece` garantit que vous n'essaierez pas de faire quelque chose de mal avec.
```

Cependant, comme vous devez effectuer une action sur ce field, vous avez besoin d'une nouvelle méthode `addTo()`. 
Ce n'est pas à l'objet `Location` de déterminer comment le `Piece` sera ajouté, alors changeons ça :

```Java
class Piece {
    private String representation;

    public String character() {
        return representation.substring(0, 1);
    }

    public void addTo(StringBuilder builder) {
        builder.append(character());
    }
}
```

Là encore, vous devez modifier la visibilité de votre attribut. 

> Pour rappel, [le principe *ouvert/fermé*](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) dit que les entités logicielles (classes, modules, fonctions, etc.) doivent être ouvertes pour l'extension, mais fermées pour la modification.

Reviewer 2: ```suggestion
Pour rappel, [le principe *ouvert/fermé*](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) dit que les entités logicielles (classes, modules, fonctions, etc.) doivent être ouvertes à l'extension, mais fermées à la modification.
```

> De plus, extraire le code pour obtenir le premier caractère du representation dans une nouvelle méthode semble être une bonne idée car il peut être réutilisé à un moment donné. 

Reviewer 2: ```suggestion
De plus, extraire le code pour obtenir le premier caractère de `representation` dans une nouvelle méthode semble être une bonne idée car il peut être réutilisé à un moment donné. 
```

    }
}
```

Et voilà qui respecte cette règle et qui est en plus, plus lisible !

## Règle 6 : N'abrégez pas !

La bonne question étant : **pourquoi diable voulez-vous abréger ?**

Vous pouvez répondre que c'est parce que vous écrivez le même nom encore et encore ?
Et je répondrai que cette méthode est réutilisée plusieurs fois, et qu'elle ressemble à une **duplication de code**.

Vous direz donc ensuite que le nom de la méthode est de toute façon trop long. 
Et je vous dirai que votre classe a peut-être de **multiples responsabilités**, ce qui est mauvais car cela viole le principe de [responsabilité unique](https://en.wikipedia.org/wiki/Single-responsibility_principle.

Je dis souvent que si vous ne trouvez pas un nom décent pour une classe ou une méthode, il y a probablement un problème. 
C'est une règle que j'utilise pour concevoir un logiciel en nommant des méthodes ou des attributs.

De plus si vous abrégez, la compréhension de votre code peut devenir ambiguë, ce qui n'est pas une bonne chose, même pour vous quand vous allez vous relire plusieurs semaines après avoir écrit ce code. 

N'abbrégez pas, point final. 

## Règle 7 : Ne pas avoir d'entités trop grosses

> Tout d'abord, par entité j'entends les packages, les classes, les méthodes, bref, tout ce qui peut être extensible.

Reviewer 2: Même les packages ? `java.util` comporte une bonne centaine de classes / interfaces, je ne suis pas sûr de voir où  est le problème

Author: C'est toujours mieux de séparer ce qu'on peut dans différents packages, ainsi je trouve que c'est plus maintenable

Reviewer: La règle parle des package que l'on crée nous même

## Règle 9 : N'utilisez pas de getter et de setter !

Ma règle préférée.

Il est normal d'utiliser des getters pour obtenir l'état d'un objet, tant que vous n'utilisez pas le résultat pour prendre des décisions en dehors de l'objet. 
**Toute décision basée entièrement sur l'état d'un objet doit être prise à l'intérieur de l'objet lui-même**.

C'est pourquoi les getters/setters [sont souvent considérés comme mauvais](https://stackoverflow.com/questions/565095/are-getters-and-setters-poor-design-contradictory-advice-seen). Là encore, ils violent le [principe ouvert/fermé](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle).

Prenons un exemple : 
```Java
// Game
private int score;

public void setScore(int score) {
    this.score = score;
}

public int getScore() {
    return score;
}

// Usage
game.setScore(game.getScore() + ENEMY_DESTROYED_SCORE);
```

Dans le code ci-dessus, le `getScore()` est utilisé pour prendre une décision, vous choisissez comment augmenter votre score, au lieu de laisser cette responsabilité à l'instance de `Game`.

Une meilleure solution serait de **supprimer les getters/setters**, et de fournir des méthodes qui **ont du sens**. 

> N'oubliez pas que vous devez **dire à** la classe de faire quelque chose, et vous ne devez **pas le demander**. 

Reviewer 2: Je pense que cette phrase méritrait une petite clarification

```Java
class Piece {
    public String representation;
}
```
```Java
class Board {
    public String boardRepresentation() {
        StringBuilder builder = new StringBuilder();

        for (Location loc : squares()) {
            builder.append(loc.current.representation.substring(0, 1));
        }

        return builder.toString();
    }
}
```
Il est normal d'avoir des attributs publics dans `Piece` et `Location`.
En fait, avoir une propriété `publique` ou une propriété `private` avec getter/setter est, selon les principes des objets calisthéniques, la même chose (voir règle 9).

Cependant, la méthode `boardRepresentation()` est horrible, jetez un œil à cette ligne :

`builder.append(loc.current.representation.substring(0, 1));`

Il accède à une `Location`, puis à son field `piece`, puis au field `representation` de la référence de `Piece` sur laquelle il effectue une action.
C'est loin d'être un point par ligne.

Mais, [la loi de Déméter](http://wiki.c2.com/?LawOfDemeter) vous dit de parler uniquement à vos amis, alors faisons-le :
```Java
class Location {
    private Piece current;

    public void addTo(StringBuilder builder) {
        current.addTo(builder);
    }
}
```

> Rendre privé le field `piece` garantit que vous n'essaierez pas d'intéragir avec et ainsi de déplacer de la logique qui devrait se trouver dans la classe `Location`.

Reviewer 3: ```suggestion
Rendre privé le field `piece` garantit que vous n'essaierez pas d'interagir avec et ainsi de déplacer de la logique qui devrait se trouver dans la classe `Location`.
```

> Rendre privé le field `piece` garantit que vous n'essaierez pas d'intéragir avec et ainsi de déplacer de la logique qui devrait se trouver dans la classe `Location`.

Reviewer 3: On pourrait peut être remplacer field par champ dans tout le document 🤔
J'attends vos avis.

Author: Des fois j'ai mis `field` et d'autre fois `attributs`, j'attends des avis aussi

Reviewer 2: Personnellement dans l'ordre de préférence:
- field
- attribut
- cacahuète 
- champ

Author: Je laisse comme c'est alors, je resolve

```Java
class Piece {
    private String representation;

    public String character() {
        return representation.substring(0, 1);
    }

    public void addTo(StringBuilder builder) {
        builder.append(character());
    }
}
```

Là encore, vous devez modifier la visibilité de votre attribut. 
Pour rappel, [le principe *ouvert/fermé*](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) dit que les entités logicielles (classes, modules, fonctions, etc.) doivent être ouvertes à l'extension, mais fermées à la modification.

De plus, extraire le code pour obtenir le premier caractère de `representation` dans une nouvelle méthode semble être une bonne idée car il peut être réutilisé à un moment donné. 
Enfin, voici la classe `Board` refactorisée :

```Java
class Board {
    public String boardRepresentation() {
        StringBuilder builder = new StringBuilder();

        for (Location location : squares()) {
            location.addTo(builder);
        }

        return builder.toString();
    }
}
```

Et voilà qui respecte cette règle et qui est en plus, plus lisible !

## Règle 6 : N'abrégez pas !

> La bonne question étant : **pourquoi diable voulez-vous abréger ?**

Reviewer 3: Pas certain de cette proposition, j'attends aussi vos avis
```suggestion
La bonne question étant : « **Pourquoi diable voulez-vous abréger ?** »
```

Author: Moi je valide

Reviewer 2: Je préfère la version originale

Vous pouvez répondre que c'est parce que vous écrivez le même nom encore et encore ?
Et je répondrai que cette méthode est réutilisée plusieurs fois, et qu'elle ressemble à une **duplication de code**.

Vous direz donc ensuite que le nom de la méthode est de toute façon trop long. 

> Et je vous dirai que votre classe a peut-être de **multiples responsabilités**, ce qui est mauvais car cela viole le principe de [responsabilité unique](https://en.wikipedia.org/wiki/Single-responsibility_principle.

Reviewer 3: ```suggestion
Et je vous dirai que votre classe a peut-être de **multiples responsabilités**, ce qui est mauvais car cela viole le principe de [responsabilité unique](https://en.wikipedia.org/wiki/Single-responsibility_principle).
```

Je dis souvent que si vous ne trouvez pas un nom décent pour une classe ou une méthode, il y a probablement un problème. 
C'est une règle que j'utilise pour concevoir un logiciel en nommant des méthodes ou des attributs.

De plus si vous abrégez, la compréhension de votre code peut devenir ambiguë, ce qui n'est pas une bonne chose, même pour vous quand vous allez vous relire plusieurs semaines après avoir écrit ce code. 

> N'abbrégez pas, point final. 

Reviewer 3: ```suggestion
N'abrégez pas, point final. 
```

## Règle 7 : Ne pas avoir d'entités trop grosses

Tout d'abord, par entité j'entends les packages, les classes, les méthodes, bref, tout ce qui peut être extensible.

Pas de classe sur 50 lignes et pas de package sur 10 fichiers.
Cela dépend de vous, mais je pense que vous pouvez changer le nombre de lignes de 50 à 150.

L'idée derrière cette règle est que les fichiers longs sont plus difficiles à lire, plus difficiles à comprendre et plus difficiles à gérer.
Donc plus vous avez des entités petites, plus elles sont faciles à comprendre et donc à maintenir, et donc plus vous gagnez de temps !

## Règle 8 : Pas de classe avec plus de deux attributs d'instance

Cette règle est probablement la plus difficile, mais elle favorise une cohésion élevée et une meilleure encapsulation dans votre classe.

> Il faut noter que cette règle est liée à la règle 3.

Reviewer 3: Tester en local voir si ça fonctionne
```suggestion
Il faut noter que cette règle est liée à la [règle 3](#règle-3--encapsulez-tous-les-primitifs-et-chaînes-de-caractères).
```

L'idée derrière cette règle est de distinguer deux types de classes :

> -Les classes qui **conservent** l'état d'une **seule variable** (value objects).
> -Les classes qui **coordonnent** **deux variables distinctes**.

Reviewer 3: ```suggestion
- Les classes qui **conservent** l'état d'une **seule variable** (value objects).
- Les classes qui **coordonnent** **deux variables distinctes**.
```

Deux instances comme limite est un choix arbitraire qui vous oblige à beaucoup découpler vos classes.

## Règle 9 : N'utilisez pas de getter et de setter !

> Ma règle préférée.

Reviewer 3: Je pense que ce que j'ai écrit est juste, cependant, il se peut que ce ne soit pas le cas : dans ce cas faites le moi remarquer.
```suggestion
Probablement la règle la plus importante de toute, elle favorise une bonne conception logicielle, et est liée au [principe de responsabilité unique](https://fr.wikipedia.org/wiki/Principe_de_responsabilit%C3%A9_unique).
```

Reviewer 2: Je ne suis pas sûr de voir le rapport avec le principe de `Single Responsibility`, en revanche c'est absolument lié à l'`open close principle`

Reviewer 3: Je viens de l'enlever du coup.
Est-ce convenable comme ça ?

Reviewer 2: Je pense que cela vaut la peine de brièvement mentionner le principe `open close`

Reviewer 3: Si je ne m'abuse, quelques lignes plus bas, l'open close est mentionné

Reviewer 2: Fine by me alors.
